src/index.js

```diff
@@ -51,9 +56,34 @@ const Reliable = or(
   UTP
 )
 
-const IPFS = or(
+let _IPFS = or(
   and(Reliable, base('ipfs')),
-  WebRTCStar
+  WebRTCStar,
+  base('ipfs')
+)
+
+const _Circuit = or(
+  and(_IPFS, base('p2p-circuit'), _IPFS),
+  and(_IPFS, base('p2p-circuit')),
+  and(base('p2p-circuit'), _IPFS),
+  and(Reliable, base('p2p-circuit')),
+  and(base('p2p-circuit'), Reliable),
+  base('p2p-circuit')
+)
+
+const CircuitRecursive = () => or(
+  and(_Circuit, CircuitRecursive),
+  _Circuit
+)
+
+const Circuit = CircuitRecursive()
+
+const IPFS = or(
```

Comment:
For reference on circuit addresses take a look at - https://github.com/libp2p/specs/tree/master/relay#addressing. The above seems to take care of most of the cases, but it hasn't been cleaned up, I'll try to do it once I've got everything ironed out.